### DISCUSSION TITLE
Installation for Home Assistant Core?

### DISCUSSION THREAD

**Discussion by SvenJachalke** [hsakoh/switchbot-mqtt#48] [Q&A]
Hi,
I would like to learn if it is possible to use the plugin with HA Core via a manual installation (docker, etc.)?
Unfortunately, I can just use Core on my Synology NAS. Thanks for any advice!

**Comment by hsakoh** [comment]
Basically, communication with Home Assistant is done via MQTT, 
so it can be run as a separate Docker container instead of an HAOS add-on.

Please refer to the following link:
https://github.com/hsakoh/switchbot-mqtt/wiki/HowToInstall#running-as-an-independent-docker-container

**Reply by hsakoh** [reply]
To provide a bit more explanation...

- Image
  The image is available on both GitHub Container Registry and Docker Hub.  
  Since Docker Hub is tagged with "Latest" every time, if you're not installing it as an add-on (i.e., pulling the Docker image yourself), Docker Hub might be easier for updates.

- Ports
  The Web UI for configuring devices, etc., uses port 8099. This needs to be exposed locally.  
  Port 8098 must be exposed if you are receiving webhooks directly from the internet.  
  (However, if you use Ngrok for webhook reception, you do not need to expose port 8098 as inbound webhook reception will not be necessary.)

- Volume
  Place the `options.json` file in the physical folder mapped to `/data`.  
  Additionally, the device configuration file (`switchbot_config.json`) will be saved under `/data` as a persistent storage during execution.

**Reply by hsakoh** [reply]
Additionally, although it is not explained in detail here, you need to set up an MQTT broker (such as Mosquitto) by some method.
You also need to set up the MQTT integration in Home Assistant.
(This is mentioned in step 1, "Preparation," of the HA-OS instructions, and it remains the same even if you run it in a separate Docker container instead of as an add-on.)

**Reply by SvenJachalke** [reply]
Ah nice! MQTT Broker (Mosquito) is already running and I already use the MQTT HA integration. Getting the docker container running and placing the settings was unclear ... I missed your section about that, so thanks for the link! I will give it a try and let you know!

**Comment by SvenJachalke** [comment]
Alright, the installation worked so far that I have the docker image running!
I copied the "options.json" to the data folder and configured everything (including all API keys from switchbot and ngrok). The program successfully fetches my switchbot devices, all checkmarks are set. Also the 3 services are started successfully. Even a test of some commands with the Test button gives a "success" back. Everything seems to work, by the devices do not appear in MQTT. 

> "LogLevel": "Trace",
>   "AutoStartServices": true,
>   "Mqtt": {
>     "AutoConfig": false,
>     "Host": "<IP of my NAS>",
>     "Port": "1883",
>     "Id": "",
>     "Pw": "",
>     "Tls": false

The parenthesis seems to be necessary for the IP but not for the port?
What could be wrong?

**Reply by hsakoh** [reply]
The fact that you can run the command test indicates that the API key and secret are correct.

If the MQTT core service has started, the MQTT connection should normally be established. 
However, there might be an issue causing the MQTT connection to fail.

* Can you check the Docker logs? Since the log level is set to "Trace," the logs during the MqttCore service startup should provide fairly detailed information.
* Can you connect to the MQTT broker through another method to check if SwitchBotMqtt messages are being published?

The "Mqtt" section in the options.json file maps to the following class.
https://github.com/hsakoh/switchbot-mqtt/blob/main/src/HomeAssistant/HomeAssistantAddOn.Mqtt/MqttOptions.cs#L8
Since "Port" is mapped to an int type, please make sure to represent it as a numeric value in the JSON.
(Since it is JSON, not YAML, the value specification is strict. Double quotes("") are not needed for Port because it is an integer. For Host, which is a string, please include double quotes.)

If devices are not being detected in the MQTT integration, it is likely failing at the green line in the diagram below.
(Although it is puzzling that the service was able to start successfully...)
https://raw.githubusercontent.com/wiki/hsakoh/switchbot-mqtt/images/Architecture.png

If the device is not marked as "Enabled," the service will start successfully, but the device will not be detected in the MQTT integration.

To ensure that the settings are correct, could you share the /data/switchbot_config.json file with me? (Please make sure to hide the MAC addresses appropriately.)

**Reply by hsakoh** [reply]
One possible cause that comes to mind is if the app was started with "AutoStartServices": true before enabling the device, the MQTTDiscovery message will not be published. (If you add a device later, you need to restart each service.)

The intended setup procedure is as follows:

1. Start the app with AutoStartService: false.
1. List the devices on the device configuration page, enable them, and save.
1. Go to the services page, start each service, and check their operation.
1. Set AutoStartService: true and restart the app.

**Comment by SvenJachalke** [comment]
Ok let me sum up my findings for your really appreciated hints:
- I tested your "AutoStartService" procedure ... has not helped
- In the docker logs a could not find any suspicious about MQTT ... filtering for "MqttCore" just displays me the stopping/starting ... anything more to watch? 
- Ok the host is marked with "<IP>" and the port without "" (1883) ... that is ok
- I check with the "MQTT-Explorer" App on MacOS ... no switchbot devices/messages can be found here :/ ... so MQTT connection is the issue
- I also attached the switchbot_config.json (without the original IDs) 
[switchbot_config_.json](https://github.com/user-attachments/files/17007133/switchbot_config_.json)
I hope that helps?

**Reply by hsakoh** [reply]
* I checked the switchbot_config.json and there are no issues (the device is enabled).
* I will provide an example of the logs from my environment. [log_sample.txt](https://github.com/user-attachments/files/17007198/log_sample.txt) Are you seeing any logs like the following?
   * Requests and responses for device status, like those on lines 8-9.
   * Requests and responses for scenes, like those on lines 23-24.
   * A log indicating that the MqttCoreService has started, like on line 25.

**Reply by hsakoh** [reply]
I checked the app's behavior when it cannot connect to the MQTT broker. It is implemented to wait internally until the connection is restored and starts normally without showing any errors.

It is likely that the failure is indeed with the MQTT connection. I have just released v1.0.27 with enhanced logging for connection issues. Please try this version. You should see more detailed logs related to the connection.

To output even more detailed logs, please try setting the following environment variable:
`Logging__LogLevel__HomeAssistantAddOn`:`Trace`

**Comment by SvenJachalke** [comment]
Hi again,
sorry for my late response, but now I was able to realize you hints. 
I installed the new version of the docker image with the environment variable.
Indeed there are some new entries in the logs with show the publishing but when everything is stated some "CancellationToken" Messages appear (see end of the attached log file ... again I erased the IDs).
Does this help you?

[_switchbot-mqtt_logs.txt](https://github.com/user-attachments/files/17018803/_switchbot-mqtt_logs.txt)

**Reply by hsakoh** [reply]
Thank you for providing the logs.
After checking the logs, it seems that there is indeed a timeout issue when connecting to the MQTT broker.

It looks like we need to check the name resolution from the switchbot-mqtt container to Mosquitto, the communication path, and the firewall on the Mosquitto side.
You mentioned using a Synology NAS; are you also running Mosquitto as a Docker container?
What network mode are you using? (I'm not very familiar with Synology NAS networking, so my support may be limited.)
https://kb.synology.com/en-us/DSM/help/ContainerManager/docker_network?version=7

For example, can you enter the container and check if communication is possible using ping?

**Reply by SvenJachalke** [reply]
I think we have it! 

Ok, so the mosquitto service I ran from the official Synology package manager because I never got the docker version running properly. But the mosquitto is actually running nicely, as I also use several other docker containers for "talking to MQTT" (e.g. Zigbee2Mqtt or JaroLift_Mqtt). And all of them are running smoothly.

The Synology Container Manager is ok, but actually I use portainer to manage my images, etc. Here I see that the connected network is called "switchbot-mqtt_default" which use "bridge". 

I checked the other MQTT containers and they use "host" ... but somehow (even after shutting down the docker) I can not connect the container to host (?!)
Message: Unable to connect container: container cannot be disconnected from host network or connected to host network
(What a nice error message :D)

So I recreated the container with "host" as network settings ... and tada ... the container starts and the MQTT devices appear in Home Assistant! Great!

So I would suggest, you could update the docker compose command with the network settings for "host"?

Again, really thank you for intense help here! I appreciate that a lot! 
Greetings,
Sven
